sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -610,13 +625,18 @@ pi_result piDevicesGet(pi_platform Platform, pi_device_type DeviceType,
                        pi_uint32 *NumDevices) {
 
   assert(Platform);
+  std::lock_guard<std::mutex> Lock(Platform->DeviceCacheMutex);
   ze_driver_handle_t ZeDriver = Platform->ZeDriver;
 
   // Get number of devices supporting Level Zero
-  uint32_t ZeDeviceCount = 0;
+  uint32_t ZeDeviceCount = Platform->PiDevicesCache.size();
```

Comment:
again, if some entries in the cache were removed in `piDeviceRelease` then this is not going to see/return all the needed devices, will it?

Reply:
👍  Good catch. Right. If a device is invalidated, only partial devices will be returned from the cache, which is bad. 
It seems we should remove the call to destroy ze command list, and invalidation.
Or I have to maintain a flag that indicates the need to get devices from L0 RT afresh.

Reply:
Done